File: src/agent/llm/anthropic.py
"""Anthropic Claude LLM provider."""

from __future__ import annotations

import json
from typing import Any, AsyncIterator

import anthropic

from agent.config import AnthropicConfig, ReasoningConfig
from agent.llm.base import (
    ResponseComplete,
    StreamEvent,
    TextDelta,
    ThinkingDelta,
    ToolUseEvent,
    UsageInfo,
)


class AnthropicProvider:
    """Wraps the Anthropic SDK for streaming chat with tool use."""

    def __init__(self, config: AnthropicConfig, model: str) -> None:
        self._model = model
        kwargs: dict[str, Any] = {}
        if config.api_key:
            kwargs["api_key"] = config.api_key
        self._client = anthropic.AsyncAnthropic(**kwargs)

    async def stream(
        self,
        system: str,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
        max_tokens: int = 8192,
        reasoning: ReasoningConfig | None = None,
    ) -> AsyncIterator[StreamEvent]:
        kwargs: dict[str, Any] = {
            "model": self._model,
            "max_tokens": max_tokens,
            "messages": messages,
        }
        if system:
            kwargs["system"] = system
        if tools:
            kwargs["tools"] = tools

        # Adaptive thinking
        if reasoning and reasoning.enabled:
            kwargs["thinking"] = {"type": "adaptive"}
            effort = reasoning.effort if reasoning.effort in ("low", "medium", "high") else "high"
            kwargs["output_config"] = {"effort": effort}

        async with self._client.messages.stream(**kwargs) as stream:
            current_block_type = ""
            current_tool_id = ""
            current_tool_name = ""
            tool_input_json = ""

            async for event in stream:
                if event.type == "content_block_start":
                    block = event.content_block
                    if block.type == "tool_use":
                        current_block_type = "tool_use"
                        current_tool_id = block.id
                        current_tool_name = block.name
                        tool_input_json = ""
                    elif block.type == "thinking":
                        current_block_type = "thinking"
                    elif block.type == "text":
                        current_block_type = "text"

                elif event.type == "content_block_delta":
                    delta = event.delta
                    if delta.type == "text_delta":
                        yield TextDelta(text=delta.text)
                    elif delta.type == "thinking_delta":
                        yield ThinkingDelta(text=delta.thinking)
                    elif delta.type == "input_json_delta":
                        tool_input_json += delta.partial_json

                elif event.type == "content_block_stop":
                    if current_block_type == "tool_use" and current_tool_name:
                        try:
                            input_data = json.loads(tool_input_json) if tool_input_json else {}
                        except json.JSONDecodeError:
                            input_data = {}
                        yield ToolUseEvent(
                            id=current_tool_id,
                            name=current_tool_name,
                            input=input_data,
                        )
                    current_block_type = ""
                    current_tool_id = ""
                    current_tool_name = ""
                    tool_input_json = ""

                elif event.type == "message_stop":
                    pass

            # Get final message for usage and thinking blocks (with signatures)
            final = await stream.get_final_message()
            thinking_blocks = []
            for block in final.content:
                if block.type == "thinking":
                    thinking_blocks.append({
                        "type": "thinking",
                        "thinking": block.thinking,
                        "signature": block.signature,
                    })
            yield ResponseComplete(
                usage=UsageInfo(
                    input_tokens=final.usage.input_tokens,
                    output_tokens=final.usage.output_tokens,
                ),
                stop_reason=final.stop_reason or "",
                thinking_blocks=thinking_blocks,
            )

File: src/llm/anthropic.ts
/** Anthropic Claude LLM provider. */

import Anthropic from '@anthropic-ai/sdk';

import type { AnthropicConfig } from '../config/schema.js';
import type { ThinkingBlock } from '../core/message.js';
import {
  responseComplete,
  textDelta,
  thinkingDelta,
  toolUseEvent,
  type LLMProvider,
  type StreamEvent,
  type StreamParams,
} from './base.js';

// Effort levels accepted by the Anthropic adaptive-thinking output_config.
// "auto" is a sentinel — when selected we omit output_config so the API picks
// its own adaptive level rather than us biasing it.
export const VALID_EFFORTS = ['low', 'medium', 'high', 'xhigh', 'max', 'auto'] as const;

// Shortcuts for the /model command — maps aliases to full Anthropic model IDs.
export const MODEL_ALIASES: Record<string, string> = {
  opus: 'claude-opus-4-7',
  sonnet: 'claude-sonnet-4-6',
  haiku: 'claude-haiku-4-5-20251001',
};

export class AnthropicProvider implements LLMProvider {
  model: string;
  private client: Anthropic;

  constructor(config: AnthropicConfig, model: string) {
    this.model = model;
    this.client = new Anthropic(config.api_key ? { apiKey: config.api_key } : {});
  }

  async *stream(params: StreamParams): AsyncIterable<StreamEvent> {
    const { system, messages, tools, maxTokens = 8192, reasoning } = params;

    const kwargs: Record<string, any> = {
      model: this.model,
      max_tokens: maxTokens,
      messages,
    };
    if (system) kwargs.system = system;
    if (tools && tools.length > 0) kwargs.tools = tools;

    // Adaptive thinking. On Opus 4.7+, thinking.display defaults to "omitted"
    // (signature-only); we ask for summarized text whenever thinking is surfaced.
    if (reasoning && reasoning.enabled) {
      const display = reasoning.show_thinking ? 'summarized' : 'omitted';
      kwargs.thinking = { type: 'adaptive', display };
      const effort = (VALID_EFFORTS as readonly string[]).includes(reasoning.effort)
        ? reasoning.effort
        : 'high';
      if (effort !== 'auto') {
        kwargs.output_config = { effort };
      }
    }

    const stream = this.client.messages.stream(kwargs as any);

    let currentBlockType = '';
    let currentToolId = '';
    let currentToolName = '';
    let toolInputJson = '';

    for await (const event of stream) {
      if (event.type === 'content_block_start') {
        const block = event.content_block;
        if (block.type === 'tool_use') {
          currentBlockType = 'tool_use';
          currentToolId = block.id;
          currentToolName = block.name;
          toolInputJson = '';
        } else if (block.type === 'thinking') {
          currentBlockType = 'thinking';
        } else if (block.type === 'text') {
          currentBlockType = 'text';
        }
      } else if (event.type === 'content_block_delta') {
        const delta = event.delta as any;
        if (delta.type === 'text_delta') {
          yield textDelta(delta.text);
        } else if (delta.type === 'thinking_delta') {
          yield thinkingDelta(delta.thinking);
        } else if (delta.type === 'input_json_delta') {
          toolInputJson += delta.partial_json;
        }
      } else if (event.type === 'content_block_stop') {
        if (currentBlockType === 'tool_use' && currentToolName) {
          let input: Record<string, any> = {};
          if (toolInputJson) {
            try {
              input = JSON.parse(toolInputJson);
            } catch {
              input = {};
            }
          }
          yield toolUseEvent(currentToolId, currentToolName, input);
        }
        currentBlockType = '';
        currentToolId = '';
        currentToolName = '';
        toolInputJson = '';
      }
    }

    // Final message for usage and thinking blocks (with signatures).
    const final = await stream.finalMessage();
    const thinkingBlocks: ThinkingBlock[] = [];
    for (const block of final.content) {
      if (block.type === 'thinking') {
        thinkingBlocks.push({
          type: 'thinking',
          thinking: block.thinking,
          signature: block.signature,
        });
      }
    }
    yield responseComplete(
      {
        input_tokens: final.usage.input_tokens,
        output_tokens: final.usage.output_tokens,
      },
      final.stop_reason ?? '',
      thinkingBlocks,
    );
  }
}
